src/Ztm.WebApi/SqlCallbackRepository.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Net;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Ztm.Data.Entity.Contexts;
+using Ztm.Data.Entity.Contexts.Main;
+
+namespace Ztm.WebApi
+{
+    public class SqlCallbackRepository : ICallbackRepository
+    {
+        readonly IMainDatabaseFactory db;
+
+        public SqlCallbackRepository(IMainDatabaseFactory db)
+        {
+            this.db = db;
+        }
+
+        public async Task<Callback> AddAsync(IPAddress ip, DateTime requestTime, Uri url, CancellationToken cancellationToken)
+        {
+            if (ip == null)
+            {
+                throw new ArgumentNullException(nameof(ip));
+            }
+
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+
+            var callback = new Callback(Guid.NewGuid(), ip, requestTime, false, url);
+
+            using (var db = this.db.CreateDbContext())
+            using (var dbtx = await db.Database.BeginTransactionAsync(IsolationLevel.Serializable, cancellationToken))
+            {
+                await db.WebApiCallbacks.AddAsync(ToEntity(callback));
+                await db.SaveChangesAsync(cancellationToken);
+
+                var webApiCallback = await db.WebApiCallbacks.FirstOrDefaultAsync(c => c.Id == callback.Id);
```

Comment:
We don't need to re-query the database, just return `callback` should be better.